### PR TITLE
#2272 Split LFRicFields into domain/lfric module

### LINKED ISSUE #2272: Splitting LFRicFields from src/psyclone/dynamo0p3.py
Sub-issue of #2235

### PR DESCRIPTION
**Pull request by oakleybrunt**
Closes #2272

### CODE REVIEW

**Comment by arporter** [comment]
Hi @oakleybrunt, is this the next one you'd like me to review? If so, please could you bring up to master and fix the conflicts. Thanks :-)

**Comment by TeranIvy** [comment]
@arporter, @oakleybrunt, I'm happy to review this. @oakleybrunt, could you please bring the branch up to date and resolve conflicts?

**Comment by oakleybrunt** [comment]
@TeranIvy Can you help me understand why the sphinx-link-check is failing please?

**Comment by arporter** [comment]
> @TeranIvy Can you help me understand why the sphinx-link-check is failing please?

It's not your fault (or problem) so you can safely ignore it :-)

File: src/psyclone/domain/lfric/lfric_kern.py [hunk 1]
```diff
@@ -609,13 +609,13 @@ def gen_stub(self):
...
         # pylint: disable=import-outside-toplevel
-        from psyclone.domain.lfric import LFRicScalarArgs
+        from psyclone.domain.lfric import LFRicScalarArgs, LFRicFields
         from psyclone.dynamo0p3 import (DynCellIterators, DynDofmaps,
                                         DynFunctionSpaces, DynCMAOperators,
                                         DynBoundaryConditions,
                                         DynLMAOperators, LFRicMeshProperties,
-                                        DynBasisFunctions, LFRicFields,
-                                        DynReferenceElement, DynStencils)
+                                        DynBasisFunctions, DynStencils,
```
Comment (TeranIvy) [inline src/psyclone/domain/lfric/lfric_kern.py:617]: Would you mind fixing a couple of `pylint` errors below (not caused by you)? 
```
src/psyclone/domain/lfric/lfric_kern.py:450:0: C0325: Unnecessary parens after 'if' keyword (superfluous-parens)
src/psyclone/domain/lfric/lfric_kern.py:715:8: R1725: Consider using Python 3 style super() without arguments (super-with-arguments)
```

File: src/psyclone/tests/domain/lfric/lfric_field_mdata_test.py [hunk 2]
```diff
@@ -35,6 +35,8 @@
 #         I. Kavcic, A. Coughtrie and L. Turner, Met Office;
 #         C. M. Maynard, Met Office/University of Reading;
 #         J. Henrichs, Bureau of Meteorology.
+#
+# Modified: O. Brunt, Met Office
```
Comment (TeranIvy) [inline src/psyclone/tests/domain/lfric/lfric_field_mdata_test.py:39]: I'd suggest adding this with other MO contributors.

File: src/psyclone/tests/domain/lfric/lfric_field_mdata_test.py [hunk 3]
```diff
@@ -46,8 +48,8 @@
 import fparser
 from fparser import api as fpapi
 from psyclone.core.access_type import AccessType
-from psyclone.domain.lfric import LFRicArgDescriptor, LFRicConstants
-from psyclone.dynamo0p3 import DynKernMetadata, LFRicFields
+from psyclone.domain.lfric import LFRicArgDescriptor, LFRicConstants, LFRicFields
```
Comment (TeranIvy) [inline src/psyclone/tests/domain/lfric/lfric_field_mdata_test.py:51]: `pycodestyle` reports `domain/lfric/lfric_field_mdata_test.py:51:80: E501 line too long (81 > 79 characters)`

**Comment by TeranIvy** [comment]
Nice work, @oakleybrunt, and apologies for the delay in reviewing!

There are some minor style issues to address (see inline comments). Tests with compilaton pass and `lfric` examples run. There are no changes to documentation.

After the review issues are addressed I would suggest updating the branch with the master and running `pycodestyle` over the entire source and tutorials (e.g. `pycodestyle src/psyclone`, `pycodestyle tutorials` from the top-level directory). I unknowingly picked up some errors when I updated a recent PR branch, it's worth checking it!

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Nice work, @oakleybrunt, and apologies for the delay in reviewing!

There are some minor style issues to address (see inline comments). Tests with compilaton pass and `lfric` examples run. There are no changes to documentation.

After the review issues are addressed I would suggest updating the branch with the master and running `pycodestyle` over the entire source and tutorials (e.g. `pycodestyle src/psyclone`, `pycodestyle tutorials` from the top-level directory). I unknowingly picked up some errors when I updated a recent PR branch, it's worth checking this!

File: src/psyclone/domain/lfric/lfric_fields.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                    DeclGen(parent, datatype=fld_dtype, kind=fld_kind,
+                            intent=fld.intent,
+                            dimension=undf_name,
+                            entity_decls=[fld.name + "_" +
+                                          fld.function_space.mangled_name]))
+
+
+# ---------- Documentation utils -------------------------------------------- #
+# The list of module members that we wish AutoAPI to generate
+# documentation for. (See https://psyclone-ref.readthedocs.io)
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/domain/lfric/lfric_fields.py]: Not quite there, but fixed now.

**Review by TeranIvy — APPROVED** [review]
All well now. There is apparent lack of coverage after moving `LFRicFields`, however further check in `dynamo0p3.py` does not show missing lines. Also, there is a link check that times out, however the link exists so I will consider it a fluke.

Tests run and documentation builds. Updated changelog and user guide for merge to master. Proceeding to merge.

**Pull request merged**